exec/java-exec/src/main/java/org/apache/drill/exec/work/batch/UnlimitedRawBatchBuffer.java

```diff
@@ -90,14 +100,39 @@ public boolean isEmpty() {
 
     @Override
     public void add(RawFragmentBatch batch) {
+      int recordCount = batch.getHeader().getDef().getRecordCount();
+      long bathByteSize = batch.getByteCount();
+      if (recordCount != 0) {
+        //skip first header batch
+        totalBatchSize += bathByteSize;
+        sampleTimes++;
+      }
+      if (sampleTimes == maxSampleTimes) {
+        long averageBathSize = totalBatchSize / sampleTimes;
+        //make a decision
+        long limit = context.getAllocator().getLimit();
```

Comment:
Another issue is the question of how many of these receivers exist per Drillbit. I don't know the answer. If I have 5 minor fragments on this Drillbit, will all 5 have their own flow control calcs? Will I have 5 fragments each trying to use 50% of 10GB for a total of 20GB of buffering? Will this be a problem?

Also, how can this algorithm go wrong? Suppose I have a set of files organized by time. I do a time range query. The first few batches might have very few rows because the filter is picking up just a few early arrivals. We see three batches, say, where the filter had low selectivity, of a few dozen rows, then decide we can hold many batches.

Later, the scan hits the bulk of my time ranges and the batches have far fewer rows filtered out. Suddenly, we need far more memory for these low-selectivity batches.

Do we need a safety valve that says that we will back off if we suddenly see large batches?

Reply:
A good valuable question!  thanks, but still a tough question. It is a resource management problem.  Firstly, I did not know the memory limit using a default 10 gb value until you mentioned it. Thanks for pointing out this. 

To drill's random statistic based memory allocation mechanism, it could not manage one drillbit's memory allocation accurately. Whatever we do ,we will risk OOM.  Think about what other system do, they will account for the total memory used by one node, if the node has not enough memory to satisfy the query, the resource manager will not assign a query to that node. After a node has accepted a query, they normally specify the network memory space and the execution memory space size separately.  So the system would be memory resource safety. 

To solve the memory risk not only to this PR but also to current implementation, maybe we need a centralized resource manager not current peer node sketchy resource management model.

We could specify a configure item to set the max network memory usage to an appropriate value (according to our cluster hardware memory size )to a fragment node. But we could not control the parallel receiver nodes happened at one drillbit. 

To your last question, the sender's current implementation will hold a calculated number rows to flush out a batch. So the row number of batch would not be a problem.  But to some situation, like the row has a text column, it's possible that the text size will not be uniform at different batches, it's possible to OOM. 

Maybe we could check for OOM .If that happened, we could notify the sender to change to a half credit value ?

Reply:
Or rollback to the initial 3 credit.

Reply:
One more answer to this question, to current SpoolingRawBatchBuffer, though it could spill to disk, but when to spill is still not memory safe. The reason is still we could not control the parallel receiver nodes one drillbit.  

I also find a bug to its implementation. It calculates the memory usage of one batch by calling the `RawFragmentBatchWrapper.getBodySize()` .  That method would always return 0. As the Drillbuf's writerIndex and readerIndex are all zero at that time.

Reply:
As it turns out, it is VERY hard to compute the memory used for a batch. The "batch sizer" classes use a mechanism that dives below the level of vectors into the ledgers. See `ValueVector.void collectLedgers(Set<BufferLedger> ledgers)`. This turned out to be the only solution because deserialized data shares buffers, as I recall, and counting buffer sizes gives the wrong answer. (It's been a couple of years since we we added that code, so I can't remember the details.)

There are two better solutions. One would be to carry the data size along with each batch. But, even that has the problems described in my earlier note.

The best solution would be to have an RM solution set the target batch size. Each operator would work to produce batches of that size. (The scan can do so using the Result Set Loader. Other operators can use the "batch sizer" method.) That way memory calcs are easier not just for this feature, but also for sorts and other buffering operators.

Since adding a target batch size is beyond the scope of this PR, I'd suggest adding some config options instead:

* Enable/disable this feature
* Target memory limit for buffering

That way, until we can control batch sizes, we can at least adjust the buffering here to deal with different workloads.

It is probably fine to make these config (boot) options rather than system options.

Finally, my (limited) understanding is that spooling for the receiver exists, but has never been used. First, it is very complex. Second, it could actually slow query execution because of the need to write to, then read from disk. Third, we'd need a way to manage maximum disk space used for spooling.

So, given that the code has never been used (I believe), it is not surprising that it contains bugs. Thanks for finding them.

Last I checked (my knowledge may be old), only sort, join and hash agg can spill to disk well.